src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -0,0 +1,414 @@
+
+                if ((tupleConversion.Conversion.Kind == ConversionKind.ImplicitTupleLiteral || tupleConversion.Conversion.Kind == ConversionKind.Identity)
+                    && IsTupleExpression(tupleConversion.Operand.Kind))
+                {
+                    return false;
+                }
+            }
+
+            Debug.Assert(tuple.Type?.IsTupleType == true);
+            return true;
+        }
+
+        // The tuple itself was already saved to temp if necessary
+        // If getting the element of a tuple literal corresponding to a single operator, we save it into initialization side-effects
+        // If getting the element of a tuple literal corresponding to a nested operator, we just return it
+        // If getting the element of a tuple type, we save it into the local side-effects
+        private BoundExpression GetTuplePart(BoundExpression tuple, int i, ImmutableArray<TupleBinaryOperatorInfo> operators,
+            ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> initEffects)
+        {
+            // Example:
+            // (1, 2) == (1, 2);
+            if (IsTupleExpression(tuple.Kind))
+            {
+                return MakeTemp(((BoundTupleExpression)tuple).Arguments[i], operators[i], temps, initEffects);
+            }
+
+            // Example:
+            // (1L, 2L == (1, 2);
+            // (1, "hello") == (1, null);
+            if (tuple.Kind == BoundKind.Conversion)
+            {
+                var tupleConversion = (BoundConversion)tuple;
+                if (tupleConversion.Conversion.Kind == ConversionKind.ImplicitNullable)
+                {
+                    return GetTuplePart(tupleConversion.Operand, i, operators, temps, initEffects);
+                }
+
+                if ((tupleConversion.Conversion.Kind == ConversionKind.ImplicitTupleLiteral || tupleConversion.Conversion.Kind == ConversionKind.Identity)
+                    && IsTupleExpression(tupleConversion.Operand.Kind))
+                {
+                    var coreArgument = WithoutImplicitNullableConversions(((BoundTupleExpression)tupleConversion.Operand).Arguments[i]);
+                    return MakeTemp(coreArgument, operators[i], temps, initEffects);
+                }
+            }
+
+            Debug.Assert(tuple.Type.IsTupleType);
+
+            // Example:
+            // t == GetTuple();
+            // t == ((byte, byte)) (1, 2);
+            // t == ((short, short))((int, int))(1L, 2L);
+            return MakeTupleFieldAccessAndReportUseSiteDiagnostics(tuple, tuple.Syntax, tuple.Type.TupleElements[i]);
+        }
+
+        private BoundExpression MakeTemp(BoundExpression expr, TupleBinaryOperatorInfo op, ArrayBuilder<LocalSymbol> temps, ArrayBuilder<BoundExpression> initEffects)
+        {
+            Debug.Assert(temps != null);
+
+            if (op.IsSingle())
+            {
+                return EvaluateSideEffectingArgumentToTemp(VisitExpression(expr), initEffects, ref temps);
```

Comment:
>return EvaluateSideEffectingArgumentToTemp(VisitExpression(expr), initEffects, ref temps); [](start = 16, length = 90)

I find it very hard to reason about when and what is getting visited and basing the logic of off TupleBinaryOperatorInfo kind makes it worse, I think.